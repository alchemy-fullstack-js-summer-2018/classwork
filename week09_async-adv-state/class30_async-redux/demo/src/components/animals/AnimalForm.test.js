import AnimalForm from './AnimalForm';
import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

describe('Animal Form', () => {
  
  it('renders add if no animal prop', () => {
    const handleComplete = jest.fn();
    const promise = Promise.resolve();
    handleComplete.mockReturnValueOnce(promise);

    const wrapper = mount(<AnimalForm onComplete={handleComplete}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();

    const animal = {
      name: 'felix',
      type: 'cat'
    };

    wrapper.find('input[name="name"]').simulate('change', {
      target: { 
        name: 'name',
        value: animal.name
      }
    });
    
    wrapper.find('input[name="type"]').simulate('change', {
      target: { 
        name: 'type',
        value: animal.type 
      }
    });

    wrapper.find('button').simulate('submit');

    // test we got the search term
    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1); // only called once
    expect(calls[0][0]).toEqual(animal);

    expect(toJSON(wrapper)).toMatchSnapshot();

    return promise
      .then(() => {
        wrapper.update();
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
  });
  
  it('renders edit if is animal prop', () => {
    const handleComplete = jest.fn();
    const promise = Promise.resolve();
    handleComplete.mockReturnValueOnce(promise);
    const handleCancel = jest.fn();

    const animal = { key: 'abc', name: 'felix', type: 'calico' };

    const wrapper = mount(<AnimalForm 
      onComplete={handleComplete}
      onCancel={handleCancel}
      animal={animal}
    />);

    expect(toJSON(wrapper)).toMatchSnapshot();

    wrapper.find('input[name="type"]').simulate('change', {
      target: { 
        name: 'type',
        value: 'tuxedo' 
      }
    });

    wrapper.find('button[type="submit"]').simulate('submit');

    // test we got the search term
    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1); // only called once
    expect(calls[0][0]).toEqual({
      ...animal,
      type: 'tuxedo'
    });

    return promise
      .then(() => {
        wrapper.update();
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    
    // wrapper.find('button[type="button"]').simulate('click');
    // expect(handleCancel.mock.calls.length).toBe(1);

    

  });
});
