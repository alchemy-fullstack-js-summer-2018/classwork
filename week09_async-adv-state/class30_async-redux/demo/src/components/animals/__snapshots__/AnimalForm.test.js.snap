// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animal Form renders add if no animal prop 1`] = `
<AnimalForm
  onComplete={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="type"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          type
          :
          <input
            name="type"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</AnimalForm>
`;

exports[`Animal Form renders add if no animal prop 2`] = `
<AnimalForm
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "name": "felix",
            "type": "cat",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value="felix"
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value="felix"
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="type"
      onChange={[Function]}
      value="cat"
    >
      <p>
        <label>
          type
          :
          <input
            name="type"
            onChange={[Function]}
            required={true}
            value="cat"
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</AnimalForm>
`;

exports[`Animal Form renders add if no animal prop 3`] = `
<AnimalForm
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "name": "felix",
            "type": "cat",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value="felix"
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value="felix"
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="type"
      onChange={[Function]}
      value="cat"
    >
      <p>
        <label>
          type
          :
          <input
            name="type"
            onChange={[Function]}
            required={true}
            value="cat"
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</AnimalForm>
`;

exports[`Animal Form renders edit if is animal prop 1`] = `
<AnimalForm
  animal={
    Object {
      "key": "abc",
      "name": "felix",
      "type": "calico",
    }
  }
  onCancel={[MockFunction]}
  onComplete={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value="felix"
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value="felix"
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="type"
      onChange={[Function]}
      value="calico"
    >
      <p>
        <label>
          type
          :
          <input
            name="type"
            onChange={[Function]}
            required={true}
            value="calico"
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Update
      </button>
      <button
        onClick={[MockFunction]}
        type="button"
      >
        Cancel
      </button>
    </p>
  </form>
</AnimalForm>
`;

exports[`Animal Form renders edit if is animal prop 2`] = `
<AnimalForm
  animal={
    Object {
      "key": "abc",
      "name": "felix",
      "type": "calico",
    }
  }
  onCancel={[MockFunction]}
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "key": "abc",
            "name": "felix",
            "type": "tuxedo",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="type"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          type
          :
          <input
            name="type"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Update
      </button>
      <button
        onClick={[MockFunction]}
        type="button"
      >
        Cancel
      </button>
    </p>
  </form>
</AnimalForm>
`;
